src/Algebra/Linear/Properties/VectorSpace.agda

```diff
@@ -0,0 +1,185 @@
+      ; *ᵣ-homo = λ r x → ∙-comm-*ᵣ
+      }
+  }
+
+------------------------------------------------------------------------
+-- Properties of linear maps from vectors to their underlying scalars.
+module _ (lm : LinearMap mod ⟨module⟩) where
+
+  open LinearMap lm
+  open ModHomo   mod ⟨module⟩ (LinearMap.homo lm)
+
+  vred : (V → S) → List V → S
+  vred g = foldr (_+_ ∘ uncurry _*_ ∘ < g , f >) 0#
+
+  foldr-homo : (g : V → S) → (xs : List V) → f (vgen g xs) ≈ vred g xs
+  foldr-homo g []       = 0ᴹ-homo
+  foldr-homo g (x ∷ xs) = begin⟨ setoid ⟩
+    f (h x (foldr h 0ᴹ xs))
+      ≈⟨ +ᴹ-homo (g x *ₗ x) (foldr h 0ᴹ xs) ⟩
+    f (g x *ₗ x) + f (foldr h 0ᴹ xs)
+      ≈⟨ +-congʳ (*ₗ-homo (g x) x) ⟩
+    g x * f x + f (foldr h 0ᴹ xs)
+      ≈⟨ +-congˡ (foldr-homo g xs) ⟩
+    g x * f x + vred g xs
+      ∎
+    where
+    h = _+ᴹ_ ∘ vscale g
+
+  vSum : List V → V
+  vSum xs = foldr _+ᴹ_ 0ᴹ xs
+
+  fScale : V → V
+  fScale = vscale f
+
+  fGen : List V → V
+  fGen = vgen f
+
+  f≈v∙ : ∀ {a} → f a ≈ v lm ∙ a
+  f≈v∙ {a} = sym (begin⟨ setoid ⟩
+    v′ ∙ a ≈⟨ ∙-comm ⟩
+    a ∙ v′ ≈⟨ foldr-homo-∙ (vscale-cong f ⟦⟧-cong) basisSet ⟩
+    foldr (_+_ ∘ (a ∙_) ∘ fScale) (a ∙ 0ᴹ) basisSet
+      ≈⟨ foldr-cong (λ {y≈z _ → +-congˡ y≈z}) ∙-idʳ basisSet ⟩
+    foldr (_+_ ∘ (a ∙_) ∘ (uncurry _*ₗ_) ∘ < f , id >) 0# basisSet
+      ≈⟨ foldr-cong (λ y≈z _ → +-cong ∙-comm-*ₗ y≈z)
+                    (reflexive Eq.refl) basisSet ⟩
+    foldr (_+_ ∘ (uncurry _*_) ∘ < f , (a ∙_) >) 0# basisSet
+      ≈⟨ foldr-cong (λ y≈z x → +-cong (*-comm (f x) (a ∙ x)) y≈z)
+                    (reflexive Eq.refl) basisSet ⟩
+    foldr (_+_ ∘ (uncurry _*_) ∘ < (a ∙_) , f >) 0# basisSet
+      ≈⟨ sym (foldr-homo (a ∙_) basisSet) ⟩
+    f (foldr (_+ᴹ_ ∘ (uncurry _*ₗ_) ∘ < (a ∙_) , id >) 0ᴹ basisSet)
+      ≈⟨ ⟦⟧-cong (Setoid.sym ≈ᴹ-setoid (basisComplete)) ⟩
+    f a ∎)
+    where
+    v′ = v lm
+
+  -- Inner product extensional equivalence.
+  x·z≈y·z⇒x≈y : ∀ {x y} → DoubleNegationElimination ℓm →
+                 ∃₂ (λ s z → ((s *ₗ (x +ᴹ -ᴹ y) ≈ᴹ z) × (f z ≉ 0#))) →
+                 (∀ {z} → x ∙ z ≈ y ∙ z) → x ≈ᴹ y
+  x·z≈y·z⇒x≈y {x} {y} dne Σ[z]fz≉𝟘 x∙z≈y∙z = fx≈fy⇒x≈y {dne} Σ[z]fz≉𝟘 fx≈fy
+    where
+    v′ = v lm
+    fx≈fy : f x ≈ f y
+    fx≈fy = begin⟨ setoid ⟩
+      f x   ≈⟨ f≈v∙ {x} ⟩
+      v′ ∙ x ≈⟨ ∙-comm ⟩
+      x ∙ v′ ≈⟨ x∙z≈y∙z ⟩
+      y ∙ v′ ≈⟨ ∙-comm ⟩
+      v′ ∙ y ≈⟨ sym (f≈v∙ {y}) ⟩
+      f y   ∎
+
+-- Isomorphism 1: (V ⊸ S) ↔ V
+V⊸S↔V : Inverse (≈ᴸ-setoid mod ⟨module⟩) ≈ᴹ-setoid
+V⊸S↔V = record
+  { to        = v
+  ; from      = λ u  → mkLM (u ∙_) u∙-homo
+  ; to-cong   = vgen-cong basisSet
+  ; from-cong = λ z x → ∙-congʳ z
+  ; inverse   = (λ v → begin⟨ ≈ᴹ-setoid ⟩
+                     vgen (v ∙_) basisSet
+                       ≈⟨ Setoid.sym ≈ᴹ-setoid basisComplete ⟩
+                     v ∎ )
+                , λ lm x → begin⟨ setoid ⟩
+                      v lm ∙ x ≈⟨ sym (f≈v∙ lm) ⟩
+                      LinearMap.f lm x   ∎
```

Comment:
As a first step could you expand these inline proofs out into a `where` clause?

Reply:
Changed as requested in next push.